### DISCUSSION TITLE
更换SongDataMap的键以容许cue存储到媒体库内

### DISCUSSION THREAD

**Discussion by lrisora** [zhongyang219/MusicPlayer2#392] [Ideas]
目前计划修改为以下，（只是编译通过，可能能用，还没测试）
```c++
    struct SongDataMapKey
    {
        wstring path;
        int index{};    // 当存储cue时用来保存音轨号，其他情况为0

        SongDataMapKey() : path(),index() {}
        SongDataMapKey(wstring path) :path(path),index() {}
        SongDataMapKey(wstring path, int index) :path(path), index(index) {}

        static size_t SongDataMapKey_hash(const SongDataMapKey& key)
        {
            return std::hash<wstring>()(key.path) ^ std::hash<int>()(key.index);
        }

        bool operator==(const SongDataMapKey& b) const
        {
            return path == b.path && index == b.index;
        }
    };

    using SongDataMap = std::unordered_map<SongDataMapKey, SongInfo, decltype(&SongDataMapKey::SongDataMapKey_hash)>;
```

需要更改CArchive存储数据，
必要添加存储 is_cue，start_pos
目前不太清楚is_cue是独立存储还是放入flags，
以及是把lengh换成start_pos+lengh还是使用start_pos+end_pos

@zhongyang219

**Comment by zhongyang219** [comment]
> 需要更改CArchive存储数据，
必要添加存储 is_cue，start_pos
目前不太清楚is_cue是独立存储还是放入flags，
以及是把lengh换成start_pos+lengh还是使用start_pos+end_pos

使用CArchive储存数据时，is_cue建议独立储存吧。
储存cue时length的储存就不用改吧，再增加一个start_pos就行了，end_pos可以通过start_pos+length算出来。
需要注意的是在CSongDataManager::SaveSongData函数中修改了储存数据的话需要提升一下写入数据的版本号，在CSongDataManager::LoadSongData函数中读取时根据数据版本号判断是否读取相应的数据。
